modules/3d/test/test_normal.cpp

```diff
@@ -396,12 +369,305 @@ class RgbdPlaneTest
+        {
+            Mat m = imread(cv::format(srcNormalsFilename.c_str(), i), IMREAD_UNCHANGED);
+
+            if (m.empty())
+            {
+                CV_Error(Error::StsError, "Failed to load normals data");
+            }
+
+            if (depth == CV_64F)
+            {
+                Mat c;
+                m.convertTo(c, CV_64F);
+                m = c;
+            }
+
+            srcNormalsCh[i] = m;
+        }
+        cv::merge(srcNormalsCh, srcNormals);
+
+        // Convert saved normals from [0; 1] range to [-1; 1]
+        srcNormals = srcNormals * 2.0 - 1.0;
+
+        // Data obtained from Blender scene
+        Matx33f intr(2000.f, 0.f, 960.f,
+                     0.f, 2000.f, 540.f,
+                     0.f, 0.f, 1.f);
+        Matx33d rotm64 = cv::Quatd(0.7805, 0.4835, 0.2087, 0.3369).conjugate().toRotMat3x3();
+        if (depth == CV_32F)
+        {
+            // Inverted camera rotation
+            Matx33f rotm(rotm64);
+
+            for (int y = 0; y < srcNormals.rows; y++)
+            {
+                Vec3f *nrmrow = srcNormals.ptr<Vec3f>(y);
+                for (int x = 0; x < srcNormals.cols; x++)
+                {
+                    Vec3f n = nrmrow[x];
+
+                    nrmrow[x] = rotm * n;
+                }
+            }
+        }
+        else if (depth == CV_64F)
+        {
+            // Inverted camera rotation
+            Matx33d rotm = rotm64;
+
+            for (int y = 0; y < srcNormals.rows; y++)
+            {
+                Vec3d *nrmrow = srcNormals.ptr<Vec3d>(y);
+                for (int x = 0; x < srcNormals.cols; x++)
+                {
+                    Vec3d n = nrmrow[x];
+
+                    nrmrow[x] = rotm * n;
+                }
+            }
+        }
+
+        Mat srcMask = srcDepth > 0;
+
+        float diffThreshold = 50.f;
+        if (scale)
+        {
+            srcDepth = srcDepth * 5000.0;
+            diffThreshold = 100000.f;
+        }
+
+        Mat srcCloud;
+        // The function with mask produces 1x(w*h) vector, this is not what we need
+        // depthTo3d(srcDepth, intr, srcCloud, srcMask);
+        depthTo3d(srcDepth, intr, srcCloud);
+        Scalar qnan = Scalar::all(std::numeric_limits<double>::quiet_NaN());
+        srcCloud.setTo(qnan, ~srcMask);
+        srcDepth.setTo(qnan, ~srcMask);
+
+        // For further result comparison
+        srcNormals.setTo(qnan, ~srcMask);
+
+        // TODO: test cases like above
+        Ptr<RgbdNormals> normalsComputer = RgbdNormals::create(1080, 1920, depth, intr, 5, diffThreshold, method);
+        normalsComputer->cache();
+
+        Mat dstNormals;
+        //TODO: LINEMOD can accept 16U and cloud instead of depth, fix it
+        if (normalsComputer->getMethod() == RgbdNormals::RGBD_NORMALS_METHOD_LINEMOD)
+        {
+            normalsComputer->apply(srcDepth, dstNormals);
+        }
+        else
+            normalsComputer->apply(srcCloud, dstNormals);
```

Comment:
ternary operator works here better, IMHO.

Reply:
Makes sense, but it's better to add an ability to accept both depth and cloud to all normal computers, then call both and compare the result to ensure it's the same.

Reply:
Since making depth input for all normal computers is not trivial, made another thing:
The only normal computer now that supports depth as input (not cloud) is LINEMOD, it's fed by `srcCloud` then `srcDepth` and the output is compared against a threshold.